api/paidAction/index.js

```diff
@@ -107,283 +200,286 @@ export default async function performPaidAction (actionType, args, context) {
-          : { mcredits: { decrement: cost } }
-      })
-    }
+async function performInternalAction (cost, actionType, paidAction, paymentMethod, args, context) {
+  const { me, models } = context
+  const run = async tx => {
+    const subContext = { ...context, tx }
+    await checkUser(me?.id ?? USER_ID.anon, subContext, paymentMethod === 'REWARD_SATS' ? cost : 0n, paymentMethod === 'FEE_CREDIT' ? cost : 0n)
+
+    try {
+      if (paymentMethod !== PAID_ACTION_PAYMENT_METHODS.ZERO_COST) {
+        if (cost <= 0n) throw new Error('invalid cost')
+        console.log('enough fee credits available, performing internal action')
+        if ((paymentMethod === 'REWARD_SATS' || paymentMethod === 'FEE_CREDIT')) {
+          await tx.user.update({
+            where: {
+              id: me?.id ?? USER_ID.anon
+            },
+            data: paymentMethod === 'REWARD_SATS'
+              ? { msats: { decrement: cost } }
+              : { mcredits: { decrement: cost } }
+          })
+          console.log('action paid with ' + paymentMethod)
+        } else throw new Error('invalid payment method ' + paymentMethod)
+      } else {
+        console.log('no cost paid action')
+      }
 
-    const result = await action.perform(args, context)
-    await action.onPaid?.(result, context)
+      const result = await performAction(null, paidAction, args, subContext)
+      await paidAction.onPaid?.({ ...result, invoice: null }, subContext)
 
-    return {
-      result,
-      paymentMethod
+      return {
+        result,
+        paymentMethod,
+        retriableWithWallet: false
+      }
+    } catch (e) {
+      console.error('action failed', e)
+      throw e
     }
-  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+  }
+  const result = context.tx
+    ? await run(context.tx)
+    : await models.$transaction(run, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
 
-  // run non critical side effects in the background
-  // after the transaction has been committed
-  action.nonCriticalSideEffects?.(result.result, context).catch(console.error)
+  // run non critical side effects in the background after the transaction has been committed
+  paidAction.nonCriticalSideEffects?.(result.result, context).catch(console.error)
   return result
 }
 
-async function performOptimisticAction (actionType, args, context) {
-  const { models } = context
-  const action = paidActions[actionType]
-
-  context.optimistic = true
-  const invoiceArgs = context.invoiceArgs ?? await createSNInvoice(actionType, args, context)
-
-  return await models.$transaction(async tx => {
-    context.tx = tx
-
-    const invoice = await createDbInvoice(actionType, args, context, invoiceArgs)
-
+async function performOptimisticAction (invoiceData, actionType, paidAction, args, context) {
+  const { models, actionId, me } = context
+  const performInvoicedAction = async tx => {
+    const subContext = { ...context, tx, optimistic: true }
```

Comment:
prevent accidental context leaking

Reply:
what do you mean by "leaking"? memory or something else?

Reply:
It is not a memory leak. It is a problem with the current code too, the way this context object is passed around in paidAction and then used to store internal values can lead to unexpected inner context leaking. 
eg. let's say you pass the context from a resolver to two paid actions

```js
something: async (parent, args, context) => {
    performPaidAction('ACTION1', {}, context)
    performPaidAction('ACTION2', {}, context)
}

```


action2 will get a context that has a bunch of random field from the action1 execution.

This never happens in our code, it is just to illustrate the issue, but in specific in our codebase, it could happen that one payment method fails in the `for (const paymentMethod of supportedPaymentMethods)` loop and when the next one is tried, it gets a dirty context with the left overs from the previous attempt. In our specific case it could end up having a closed or closing transaction in the tx field .
Making a shallow copy of the context like this, fixes this issue while keeping the current way of passing the context, and prevents future hard to debug issues related to context leaking.